lib/inputs/base-input.js

```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+class BaseInput {
+    constructor(type, id, params, inputDefaultValue) {
+        this.type = type;
+        this.id = id;
+        this.params = this.normalizeParams(params || {});
+
+        this.value = this.params.default || this.getInputDefaultValue();
+    }
+
+    getInputDefaultValue() {
+        throw new Error('getInputDefaultValue must be implemented');
+    }
+
+    normalizeParams(params) {
+        return params || {};
+    }
+
+    normalizeValue() {
+        throw new Error('normalizeValue must be implemented');
+    }
+
+    setValue(value) {
+        this.value = this.normalizeValue(value);
+    }
+
+    toObj() {
```

Comment:
Suggest renaming to `toJSON` (which is expected to return the value that should be serialized), or something like `valueOf`. I don't particularly like either of these though, @dmajda would prolly have a better suggestion.